src/spatialdata/_core/query/spatial_query.py

```diff
@@ -619,206 +698,180 @@ def _(
-            else:
-                ddf = PointsModel.parse(ddf, coordinates={"x": "x", "y": "y"})
-            set_transformation(ddf, transformation, target_coordinate_system)
-            new_points[points_name] = ddf
-
-    new_images = {}
-    if images:
-        for images_name, im in sdata.images.items():
-            min_x, min_y, max_x, max_y = polygon.bounds
-            cropped = bounding_box_query(
-                im,
-                min_coordinate=[min_x, min_y],
-                max_coordinate=[max_x, max_y],
-                axes=("x", "y"),
-                target_coordinate_system=target_coordinate_system,
-            )
-            new_images[images_name] = cropped
-    if labels:
-        for labels_name, l in sdata.labels.items():
-            _ = labels_name
-            _ = l
-            raise NotImplementedError(
-                "labels=True is not implemented yet. If you encounter this error please open an "
-                "issue and we will prioritize the implementation."
-            )
-
-    if filter_table and sdata.table is not None:
-        table = _filter_table_by_elements(sdata.table, {"shapes": new_shapes, "points": new_points})
-        if table is not None and len(table) == 0:
-            table = None
-    else:
-        table = sdata.table
-    return SpatialData(shapes=new_shapes, points=new_points, images=new_images, table=table)
-
-
-# this function is currently excluded from the API documentation. TODO: add it after the refactoring
+@singledispatch
 def polygon_query(
-    sdata: SpatialData,
-    polygons: Polygon | list[Polygon],
+    element: SpatialElement | SpatialData,
+    polygon: Polygon | MultiPolygon,
     target_coordinate_system: str,
     filter_table: bool = True,
     shapes: bool = True,
     points: bool = True,
     images: bool = True,
     labels: bool = True,
-) -> SpatialData:
+) -> SpatialElement | SpatialData | None:
     """
-    Query a spatial data object by a polygon, filtering shapes and points.
+    Query a SpatialData object or a SpatialElement by a polygon or multipolygon.
 
     Parameters
     ----------
-    sdata
-        The SpatialData object to query
+    element
+        The SpatialElement or SpatialData object to query.
     polygon
-        The polygon (or list of polygons) to query by
+        The polygon/multipolygon to query by.
     target_coordinate_system
-        The coordinate system of the polygon
-    shapes
-        Whether to filter shapes
-    points
-        Whether to filter points
+        The coordinate system of the polygon/multipolygon.
+    filter_table
+        If `True`, the table is filtered to only contain rows that are annotating regions
+        contained within the query polygon/multipolygon.
+    shapes [Deprecated]
+        This argument is now ignored and will be removed. Please filter the SpatialData object before calling this
+        function.
+    points [Deprecated]
```

Comment:
I don't think this is a good way to specify deprecation, I would just keep the docstrings and call the deprecation warning at functions all

Reply:
Done.